internal/build/imgsrc/dockerfile_builder.go

```diff
@@ -288,11 +293,94 @@ func (*dockerfileBuilder) Run(ctx context.Context, dockerFactory *dockerClientFa
 		Size: img.Size,
 	}
 
+	if opts.UseOverlaybd && dockerFactory.IsRemote() {
+		obdImage, err := buildOverlaybdImage(ctx, dockerFactory.appName, docker, opts)
+		if err != nil {
+			terminal.Warnf("failed to build lazy-loaded image, not using lazy-loading: %v", err)
+		} else {
+			di = *obdImage
+		}
+	}
+
 	span.SetAttributes(di.ToSpanAttributes()...)
 
 	return &di, "", nil
 }
 
+func buildOverlaybdImage(ctx context.Context, appName string, docker *dockerclient.Client, opts ImageOptions) (*DeploymentImage, error) {
+	if !opts.Publish {
+		return nil, errors.New("lazy loaded images require --push")
+	}
+
+	terminal.Info("Building lazy-loading image, please wait...")
+
+	daemonHost := docker.DaemonHost()
+	parsed, err := url.Parse(daemonHost)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse daemon host: %w", err)
+	}
+	hostPort := parsed.Host
+	host, _, _ := net.SplitHostPort(hostPort)
+	parsed.Host = host + ":8080"
+	parsed.Scheme = "http"
+	parsed.Path = "/flyio/v1/buildOverlaybdImage"
+	rchabUrl := parsed.String()
+
+	terminal.Debugf("rchab url: %s", rchabUrl)
+
+	repo := strings.Split(opts.Tag, ":")[0]
+	version := strings.Split(opts.Tag, ":")[1]
+
+	if !strings.HasPrefix(repo, "registry.fly.io/") {
+		return nil, fmt.Errorf("lazy loaded images must be pushed to registry.fly.io, not %s", repo)
+	}
+
+	terminal.Debugf("overlaybd repo: %s, version: %s", repo, version)
+
+	creds := registryAuth(config.Tokens(ctx).Docker())
+
+	body := map[string]string{
+		"repo":   repo,
+		"input":  version,
+		"output": version + "-obd",
+		"creds":  creds.Username + ":" + creds.Password,
+	}
+	jsonBody, err := json.Marshal(body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", rchabUrl, bytes.NewBuffer(jsonBody))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("User-Agent", fmt.Sprintf("flyctl/%s", buildinfo.Version().String()))
+	req.SetBasicAuth(appName, config.Tokens(ctx).Docker())
+
+	res, err := docker.HTTPClient().Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to post to /buildOverlaybdImage: %w", err)
+	}
+	defer res.Body.Close()
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(res.Body)
+	terminal.Debugf("rchab response: %s", buf.String())
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("building lazy image returned status %d: %s", res.StatusCode, buf.String())
+	}
+	hash := buf.String()
+
+	terminal.Info("Lazy-loading image built successfully!")
+
+	return &DeploymentImage{
+		ID:   hash,
+		Tag:  repo + ":" + version + "-obd",
+		Size: 0,
```

Comment:
Would it make sense to set the `Size` to that of the original image since (I think) it's printed to output in a few places? (I suspect the original image size is still pretty close to the overlaybd image, though I could be wrong?)

Reply:
Oh, forgot about this. The image sizes can vary by over 30%. I'd rather have a 0 here than a closer wrong number, to make it easier to detect issues. 

I can still deploy an overlaybd image without setting the size here, so I'm inclined not to worry about it at the moment.